src/Common/ZooKeeper/ZooKeeper.cpp

```diff
@@ -55,70 +58,175 @@ static void check(Coordination::Error code, const std::string & path)
-            }
-            catch (const Poco::Net::HostNotFoundException & e)
-            {
-                /// Most likely it's misconfiguration and wrong hostname was specified
-                LOG_ERROR(log, "Cannot use ZooKeeper host {}, reason: {}", host_string, e.displayText());
-            }
-            catch (const Poco::Net::DNSException & e)
-            {
-                /// Most likely DNS is not available now
-                dns_error = true;
-                LOG_ERROR(log, "Cannot use ZooKeeper host {} due to DNS error: {}", host_string, e.displayText());
-            }
+
+void ZooKeeper::updateAvailabilityZones()
+{
+    std::vector<ShuffleHost> shuffled_hosts = shuffleHosts();
+    Coordination::ZooKeeper::Nodes nodes = hostsToNodes(log, shuffled_hosts);
+
+    for (const auto & node : nodes)
+    {
+        try
+        {
+            Coordination::ZooKeeper::Nodes single_node{node};
+            auto tmp_impl = std::make_unique<Coordination::ZooKeeper>(single_node, args, zk_log);
+            auto idx = node.original_index;
+            availability_zones[idx] = tmp_impl->tryGetAvailabilityZone();
+            LOG_DEBUG(log, "Got availability zone for {}: {}", args.hosts[idx], availability_zones[idx]);
+        }
+        catch (...)
+        {
+            DB::tryLogCurrentException(log, "Failed to get availability zone for " + node.address.toString());
         }
+    }
+}
+
+void ZooKeeper::init(ZooKeeperArgs args_, std::unique_ptr<Coordination::IKeeper> existing_impl)
+{
+    args = std::move(args_);
+    log = getLogger("ZooKeeper");
+
+    if (existing_impl)
+    {
+        chassert(args.implementation == "zookeeper");
+        impl = std::move(existing_impl);
+        LOG_INFO(log, "Switching to connection to a more optimal node {}", impl->getConnectedHostPort());
+    }
+    else if (args.implementation == "zookeeper")
+    {
+        if (args.hosts.empty())
+            throw KeeperException::fromMessage(Coordination::Error::ZBADARGUMENTS, "No hosts passed to ZooKeeper constructor.");
 
-        if (nodes.empty())
+        chassert(args.availability_zones.size() == args.hosts.size());
+        if (availability_zones.empty())
         {
-            /// For DNS errors we throw exception with ZCONNECTIONLOSS code, so it will be considered as hardware error, not user error
-            if (dns_error)
-                throw KeeperException::fromMessage(Coordination::Error::ZCONNECTIONLOSS, "Cannot resolve any of provided ZooKeeper hosts due to DNS error");
-            else
-                throw KeeperException::fromMessage(Coordination::Error::ZCONNECTIONLOSS, "Cannot use any of provided ZooKeeper nodes");
+            /// availability_zones is empty on server startup or after config reloading
+            /// We will keep the az info when starting new sessions
+            availability_zones = args.availability_zones;
```

Comment:
so we have `availability_zones` in the <node>` section because it can be used for static configuration: in deployment people know their keeper node1 must be in us-west1-a; also for testing?

Keeper is implemented to allow statically configure its node availability zone already. https://github.com/ClickHouse/ClickHouse/pull/56715/files#diff-5d51e4a5ca8d2746a1536fc69a257de3cf13b44429eb97891d410925b875e2f1R16  so we should be able to suit that use case by always fetching az from keeper in the runtime.

but maybe this can be used for non ClickHouse keeper cases.

Reply:
Yes, this is for testing and for Apache ZooKeeper